start work on the migration guide (#12836)
apollographql/apollo-client: docs/source/migrating/apollo-client-4-migration.mdx

  - Many improvements and bug fixes in the new `LocalState` implementation
- The `Observable` implementation has been changed to `rxjs`.
- The incremental delivery implementation has been made pluggable, so in the future we can support multiple different versions of the incremental delivery protocol.
- For a full list of changes, see the [changelog](https://github.com/apollographql/apollo-client/blob/main/CHANGELOG.md).

## Installation

<Caution>

Apollo Client 4.0 is a major-version release that includes breaking changes. If you are updating an existing application to use Apollo Client 4.0, see the [changelog](https://github.com/apollographql/apollo-client/blob/main/CHANGELOG.md) for details about these changes.

</Caution>

Install Apollo Client 4 along with its peer dependencies with the following command:

```
npm install @apollo/client graphql rxjs
```

## Codemod

To ease the migration process, we have created a codemod that will automatically update your codebase to use the new imports and APIs in Apollo Client 4.

This codemod consists of the following steps:

- `imports` step:

>   - Updates imports that have moved around, e.g. changing `import { useQuery } from '@apollo/client'` to `import { useQuery } from '@apollo/client/react'`.

Author: Yeah, the idea was to keep it short because when I wrote that I didn't have the diffs yet, but once I add those it will be less of a problem :)

```sh
npx @apollo/client-codemod-migrate-3-to-4 --parser ts --extensions ts src
npx @apollo/client-codemod-migrate-3-to-4 --parser tsx --extensions tsx src
```

### Running only specific steps

You can also run the codemod with only specific steps by using the `--codemod` option. For example, to run only the `imports` and `links` steps, you can use:

```sh
npx @apollo/client-codemod-migrate-3-to-4 --codemod imports --codemod links src
```

## Updating imports

If you used the [Codemod](#running-the-codemod) to update your imports, you can skip this section.
If you did not use the codemod, you will need to manually update your imports to point at the right files.

### Move from manual CJS/ESM imports to `exports`

Apollo Client 4 now has an `exports` field in the `package.json`, which means that you previously used imports like `@apollo/client/react/index.js` or `@apollo/client/react/react.cjs`, you should now use the `@apollo/client/react` entry point.
Your bundler will be aware of the different module formats provided and be able to resolve the right one for you, based on import conditions and the `exports` field of the `package.json`.

<details>

> <summary><h3 style="display:inline">List of all changed imports</h3><br/>(click to expand)</summary>

Author: As for H3/H4:

We have
```md
## Updating imports
### Move from manual CJS/ESM imports to `exports`
### List of all changed imports
```

I'd see this as part of "Updating imports", it's not really a subsection of "Move from manual CJS/ESM imports to `exports`".

> <summary><h3 style="display:inline">List of all changed imports</h3><br/>(click to expand)</summary>

Reviewer: Fair enough. Let's leave as-is

| `@apollo/client/testing/core` | `@apollo/client/testing` |

The following entry points have been removed:
| previous entry point | reason |
| --------------------- | -------------------- |
| `@apollo/client/react/components` | The render prop components were already deprecated in Apollo Client 3.x and have been removed in 4. |
| `@apollo/client/react/hoc` | The higher order components were already deprecated in Apollo Client 3.x and have been removed in 4. |
| `@apollo/client/react/parser` | This module was an implementation detail of the render prop components and HOCs. |
| `@apollo/client/testing/experimental` | This is available as [@apollo/graphql-testing-library](https://github.com/apollographql/graphql-testing-library) |
| `@apollo/client/utilities/globals` | This was an implementation detail and has been removed. Some of the exports are now available in other entry points. |
| `@apollo/client/utilities/subscriptions/urql` | This is supported natively by urql and is no longer included. |

The following individual imports have been renamed, moved to new entry points and/or moved into namespaces:

<table>
<thead>
<tr><th>previous entry point</th><th>new entry point</th></tr>

> <tr><td>previous import name</td><td>new import name</td></tr>

Author: I've split this table into multiple `tbody`. Doesn't make a visual difference, but should make it clearer when looking at the markdown file.

<tr><td>`FulfilledPromise`</td><td>(unchanged)</td></tr>
<tr><td>`IsAny`</td><td>(unchanged)</td></tr>
<tr><td>`NoInfer`</td><td>(unchanged)</td></tr>
<tr><td>`PendingPromise`</td><td>(unchanged)</td></tr>
<tr><td>`Prettify`</td><td>(unchanged)</td></tr>
<tr><td>`Primitive`</td><td>(unchanged)</td></tr>
<tr><td>`RejectedPromise`</td><td>(unchanged)</td></tr>
<tr><td>`RemoveIndexSignature`</td><td>(unchanged)</td></tr>
<tr><td>`VariablesOption`</td><td>(unchanged)</td></tr>
<tr><th>`@apollo/client/utilities/global`</th><th>`@apollo/client/utilities/environment`</th></tr>
<tr><td>`DEV`</td><td>`__DEV__`</td></tr>
<tr><td>`__DEV__`</td><td>(unchanged)</td></tr>
<tr><th>`@apollo/client/utilities/global`</th><th>`@apollo/client/utilities/internal/globals`</th></tr>
<tr><td>`global`</td><td>(unchanged)</td></tr>
<tr><td>`maybe`</td><td>(unchanged)</td></tr>
<tr><th>`@apollo/client/utilities/global`</th><th>`@apollo/client/utilities/invariant`</th></tr>
<tr><td>`invariant`</td><td>(unchanged)</td></tr>
<tr><td>`InvariantError`</td><td>(unchanged)</td></tr>
<tr><td>`newInvariantError`</td><td>(unchanged)</td></tr>
</tbody>
</table>

</details>

## Replacing removed exports

> If you used the [Codemod](#running-the-codemod) to update your imports, all imports for removed exports have been replaced with imports from `@apollo/client/v4-migration`. You should get a TypeScript error everywhere you use a removed export, and hovering your mouse over it, you get more information about why the export was removed and what you can use instead.

Reviewer: I'm curious, what is it that causes the TypeScript error for the removed export? These are marked as deprecated so wouldn't you just see the strikethough?

Barring that I have a couple suggestions:

```suggestion
If you used the [Codemod](#running-the-codemod) to update your imports, all removed exports moved to the `@apollo/client/v4-migration` entry point. You should get a TypeScript error everywhere you use a removed export. When you hover over the export, you'll get more information about why the export was removed along with migration instructions.
```

- I shortened the part after the comma to try and reduce the number of times we. reference "import"/"export"  to avoid confusion
- I removed "your mouse" since Vim/Neovim users don't use a mouse 😜 (we just call it "hovering")
- Made it clear the information contains migration instructions

Author: > I'm curious, what is it that causes the TypeScript error for the removed export? These are marked as deprecated so wouldn't you just see the strikethough?

Everything in that file is typed as `never`, so you can't really use it.

> For a list of all removed imports and recommended actions, see [node_modules/@apollo/client/v4-migration.d.ts](https://app.unpkg.com/@apollo/client@^4.0.0-rc/files/v4-migration.d.ts)

Reviewer: I just realized, the first thing you see when you open that `v4-migration.d.ts` file is that `Removals` namespace. Would it be worth moving that to the bottom of the file so that you see the removed stuff immediately? I think we can do that without having to do another release (especially since we'll want to update this link anyways once we are out of rc). Thoughts?

Author: > especially since we'll want to update this link anyways once we are out of rc

It shouldn't be too bad if we forget - it's a semver range in the link - `^4.0.0-rc` should resolve to `^4.0.0` or `^4.0.5` later on.

> ## Update your ApolloClient creation

Reviewer: ```suggestion
## Update the `ApolloClient` constructor
```

I'd use "constructor" here instead of "creation" which ties in nicely with the intro paragraph. Since you're using ApolloClient (one word), let's also wrap in inline code tag.

Another idea for a headline might be:
```
## Update the initialization of `ApolloClient` 
```

Author: ```suggestion
## Update the initialization of `ApolloClient`
```
I like this one better :)

> Some of the constructor options of `ApolloClient` have changed in Apollo Client 4. This section will help you update your `ApolloClient` creation code to the new options.

Reviewer: ```suggestion
Several of the constructor options of `ApolloClient` have changed in Apollo Client 4. This section provides instruction on migrating to the new options when initializing your `ApolloClient` instance.
```

I tend to avoid "some" where I can 🙂. And like the header, let's maybe change it to "initialization" rather than "creation"

> ### Implicitly create a new `HttpLink`

Reviewer: ```suggestion
### Explicitly provide `HttpLink`
```

It seems most of the headlines are written from the perspective of what you want to do, rather than how it was. "Implicit" was v3, but now `link` is explicit, so I'd update it like this.

> The shorthand annotation where you could pass `uri`, `headers` or `credentials` directly into the `ApolloClient` constructor has been removed.
> Instead, you now need to create a new `HttpLink` instance and pass it to the `link` option of `ApolloClient`.

Reviewer: ```suggestion
The `uri`, `headers`, and `credentials` options used to implicitly create an `HttpLink` has been removed. You now need to create a new `HttpLink` instance and pass it to the `link` option of `ApolloClient`.
```

I'd make this a bit more terse by describing it this way.

Author: ```suggestion
The `uri`, `headers`, and `credentials` options used to implicitly create an `HttpLink` have been removed. You now need to create a new `HttpLink` instance and pass it to the `link` option of `ApolloClient`.
```

> Although it was convenient, it created a direct coupling with `HttpLink`, so even users that were not using `HttpLink` had to ship it in their bundle. This change enables you to use any link implementation you want, without having to ship `HttpLink` if you don't need it.

Reviewer: How about moving this after the code block? The first paragraph does a great job of describing _what_ changed so it feels like the example should be shown immediately. This is a great paragraph on _why_ it changed, but let's reserve that for after the code block.

And some suggestions:

```suggestion
Although the previous options were convenient, it meant all Apollo Client instances were bundled with `HttpLink`, even if it wasn't used in the link chain. This change enables you to use different terminating links without the increase in bundle size needed to include `HttpLink`. Most applications that scale moved away from these options anyways as soon as more complex link chains were needed.
```

- I think its less about "direct coupling" and more about shipping code that wasn't used.
- Replaced "link implementation" with "terminating link" since thats a bit more accurate in this context
- Added a statement about the increase in complexity which also provides reason why we moved away from these options since those options weren't really used as soon as more complex link chains were introduced.

Author: 9822f0040cc5ce42d57efbdbabb5882884574d9b

```ts
import {
  ApolloClient,
  InMemoryCache,
  HttpLink, // [!code ++]
} from "@apollo/client/core";
const client = new ApolloClient({
  // ...
  link: new HttpLink({ // [!code ++]
    uri: "https://example.com/graphql",
    credentials: "include",
    headers: { "x-custom-header": "value" },
  }), // [!code ++]
  // ...
});
```

> ### Move `name` and `version` into the `clientAwareness` option

Reviewer: ```suggestion
### Migrating client awareness options
```

Thoughts?

> 

Reviewer: Let's add a small "what" statement:

```suggestion

The `name` and `version` options, which are a part of the client awareness feature, have been moved into the `clientAwareness` option.
```

And after the code block, I'd suggest a why statement as well:

```mdx
This change reduced confusion on what the `name` and `version` options were meant for and provides better future-proofing for additional features that might be added later.
```

Reviewer: Also, is it worth adding a `<Note />` that this only works if you're using `HttpLink`/`BatchHttpLink`? Might be good to mention that other terminating links might need to add the `ClientAwarenessLink`.

Author: c7ff0dff704eca68141704e633e55ed4c87e032e

```ts
const client = new ApolloClient({
  // ...
  clientAwareness: { // [!code ++]
    name: "My Apollo Client App",
    version: "1.0.0",
  }, // [!code ++]
  // ...
});
```

> ### If using local state: add a new `LocalState` instance

Reviewer: ```suggestion
### Updating local state
```

Can we avoid "if" in the name? I'd rather put the "if" in the intro paragraph, or in a note in this section if we can. That makes it a bit more terse in the sidebar.

> If you are using `@client` fields, you should create a new `LocalState` instance and pass it to the `ApolloClient` constructor.

Reviewer: ```suggestion
When using `@client` fields, you now need to create a new `LocalState` instance and provide it as the `localState` option to the `ApolloClient` constructor.
```

Let's assume the reader is reading this because this change applies to them instead of starting each section with "if"

Author: I've learned very strictly when learning English that "when" is only to be used as a temporal specifier, and always to use `if` for conditional specifiers - so I'm probably overdoing that :)

```ts
import {
  ApolloClient,
  InMemoryCache,
  LocalState, // [!code ++]
} from "@apollo/client/core";
const client = new ApolloClient({
  // ...
  localState: new LocalState(), // [!code ++]
  // ...
});
```

Additionally, if you are using local resolvers, you should pass the `resolvers` option to the `LocalState` constructor instead of the `ApolloClient` constructor.

```ts
import {
  ApolloClient,
  InMemoryCache,
  LocalState, // [!code ++]
} from "@apollo/client/core";
const client = new ApolloClient({
  // ...
  localState: new LocalState({ // [!code ++]
    resolvers: {
      Query: {
        myField: () => "Hello World",
      },
    },
  }), // [!code ++]
  // ...
});
```

> 

Reviewer: Again, I think a why statement here is good.

```suggestion
Previously, all Apollo Client instances were bundled with local state management functionality, even if you didn't use the `@client` directive. This change means local state management is now opt-in and reduces the size of your bundle when you aren't using this feature.
```

> ### Change `connectToDevTools` to `devtools.enabled`

Reviewer: ```suggestion
### Change `connectToDevTools`
```

For terseness. I think we can save the "how" for the intro paragraph and the code example instead of trying to stuff it in the title

The `connectToDevTools` option has been replaced with a new `devtools` option that contains an `enabled` property.

```ts
const client = new ApolloClient({
  // ...
  connectToDevTools: true, // [!code --]
  devtools: { enabled: true }, // [!code ++]
  // ...
});
```

> ### Change `disableNetworkFetches` to `prioritizeCacheValues`

Reviewer: ```suggestion
### Change `disableNetworkFetches`
```

> If you are using the `disableNetworkFetches` option, note that it has been renamed to `prioritizeCacheValues`.

Reviewer: ```suggestion
The `disableNetworkFetches` option has been renamed to `prioritizeCacheValues` to better describe its behavior.
```

I think it makes sense to assume that a reader would read a particular section because its applicable to migrate. That means we can avoid "if" to start these sections and assume the option is used. Sections that aren't relevant will be skipped by the reader.

```ts
const client = new ApolloClient({
  // ...
  disableNetworkFetches: ..., // [!code --]
  prioritizeCacheValues: ..., // [!code ++]
  // ...
});
```

### Enable incremental delivery (`@defer`)

> In Apollo Client 4, you can swap out the incremental delivery protocol implementation.
> That means if you were using `@defer` previously, you should now enable the new `Defer20220824Handler` by passing it to the `incrementalHandler` option.

Reviewer: ```suggestion
The incremental delivery protocol implementation is now configurable and requires you to opt-in to use the proper format. If you currently use `@defer` in your application, you need to configure an incremental handler and provide it to the `incrementalHandler` option to the `ApolloClient` constructor.

Since Apollo Client 3 only supported an older version of the specification, initialize an instance of `Defer20220824Handler` and provide it as the `incrementalHandler` option.
```

```ts
import { Defer20220824Handler } from "@apollo/client/incremental"; // [!code ++]
const client = new ApolloClient({
  // ...
  incrementalHandler: new Defer20220824Handler(), // [!code ++]
  // ...
});
```

> Additionally, you should enable the `TypeOverrides` for the `Defer20220824Handler`.
> This will add the deferred chunk types to `ApolloLink.Result` and ensures that you handle incremental results correctly in custom links that might access `result` directly.

Reviewer: ```suggestion
To provide accurate TypeScript types, you will also need tell Apollo Client to use the types associated with the `Defer20220824Handler` in order to provide accurate types for incremental chunks, used with the `ApolloLink.Result` type. This ensures that you properly handle incremental results in your custom links that might access the result directly.
```

I used more direct language here to make it feel like a necessity (i.e. "you will also need" vs "you should").

> To do so, create a new `.d.ts` file in your project (e.g. `apollo-client.d.ts`) and add the following content:

Reviewer: ```suggestion
Create a new `.d.ts` file in your project (e.g. `apollo-client.d.ts`) and add the following content:
```

I think starting with this is ok 🙂

```ts title="apollo-client.d.ts
import { Defer20220824Handler } from "@apollo/client/incremental"; // [!code highlight]

declare module "@apollo/client" {
  export interface TypeOverrides extends Defer20220824Handler.TypeOverrides {} // [!code highlight]
}
```

<Note>

> Over time, we will introduce additional incremental handlers to support different proposals of the protocol.

Reviewer: ```suggestion
Apollo Client only provides support for the older incremental specification at this time. Future versions will introduce new incremental handlers to support more recent versions of the specification.
```

I just realized we've used "protocol" here, but maybe we should use specification? (if not, `s/specification/protocol` in my suggestion). I think adding a statement that makes it clear we don't ship newer versions of the protocol also help here.

To configure the data masking types to be the same as they were in Apollo Client 3, create a `.d.ts` file in your project with the following content:

```ts title="apollo-client.d.ts
import { GraphQLCodegenDataMasking } from "@apollo/client/masking";// [!code highlight]

declare module "@apollo/client" {
  export interface TypeOverrides// [!code highlight:2]
    extends GraphQLCodegenDataMasking.TypeOverrides {}
}
```

<Note>

You can always combine multiple `TypeOverrides` of different kinds, so you can also combine this with the `Defer20220824Handler` type overrides from the previous section.

```ts title="apollo-client.d.ts disableCopy=true
import { Defer20220824Handler } from "@apollo/client/incremental";
import { GraphQLCodegenDataMasking } from "@apollo/client/masking";// [!code ++]

declare module "@apollo/client" {
  export interface TypeOverrides extends Defer20220824Handler.TypeOverrides {}// [!code --]
  export interface TypeOverrides// [!code ++:3]
    extends Defer20220824Handler.TypeOverrides,
      GraphQLCodegenDataMasking.TypeOverrides {}
}
```

</Note>

## Core API changes

### New `notifyOnNetworkStatusChange` default value

> `notifyOnNetworkStatusChange` for `ObservableQuery` instances created by `client.watchQuery` now defaults to `true` instead of `false`.

Author: 92f528da144fdad09ee8ae8c1d0f978586ca1bf4

### Immediate `loading` status emitted by `ObservableQuery`

When you create an `ObservableQuery` with `client.watchQuery`, unless you explicitly disable `notifyOnNetworkStatusChange`, the first emitted value will now always have `loading: true` if the query is not already in the cache.

> Previously, this initial loading state would have been skipped.

Reviewer: Good call. I like that better.

## For React users

One of Apollo Client 4's focus points is to be more opinionated about how to use Apollo Client.
A big part of that is to ensure that the React hooks are more straightforward to use, for a very specific use case.
We believe that hooks should be used to synchronize data with your component, and not trigger side effects that won't synchronize with your component.
As a result, in some previous use cases, we now recommend to use the core APIs of Apollo Client directly where synchronization with a component is not intended.

> As an example, if you were using `useLazyQuery` to trigger many queries in quick succession, but were not interested in synchronizing the result with your component, you should now use `client.query` directly instead of using a hook.

Author: Yeah, that's a good call - adding the heading.

```ts disableCopy=true

> const [execute] = useLazyQuery(MY_QUERY);// [!code --]
> const client = useApolloClient();// [!code ++]
> // ...
> await execute({ variables });// [!code --]
> await client.query({ query: MY_QUERY, variables });// [!code ++]

Author: Yeah, but let's keep the `// ...` in there to show that these usually don't happen immediately after each other
```suggestion
const [execute] = useLazyQuery(MY_QUERY);// [!code --:3]
// ...
await execute({ variables });
const client = useApolloClient();// [!code ++:3]
// ...
await client.query({ query: MY_QUERY, variables });
```

```

### changes to `useLazyQuery`

> `useLazyQuery` has undergone a big rewrite to have a more predictable behaviour.

Author: I'm trying, but I grew up with 8 years of british english in school 🤣 

Yeah, makes sense :)

### (optional, but recommended) Stop using generated hooks

Apollo Client 4 comes with a lot of improvements to our hook types. If you are using generated hooks, you will not be able to benefit from these improvements, so we recommend to stop using generated hooks and use our hooks directly with `TypedDocumentNode` instead.

<Note>

The GraphQL Code Generator team is providing a codemod to help you migrate off of generated hooks - you can watch [this video](https://www.youtube.com/watch?v=GGmt0hvnQNU) that introduces it.

Note that that codemod is a separate package and not maintained by the Apollo Client team. Generally, we recommend to use a manual codegen configuration over the client preset shown in the video, as the client preset has some overlapping features with Apollo Client that do not play well together. If you choose to go this route, see [our recommended configuration for the GraphQL Code Generator Client Preset](https://www.apollographql.com/docs/deploy-preview/74d89f8b5deec991d04de44f/react/development-testing/static-typing#the-client-preset), but in the long run also consider rolling your own configuration.

</Note>

> ### New SSR api `prerenderStatic`

Author: I think we can leave it out tbh.

`prerenderStatic` can be used together with the modern [`prerender`](https://react.dev/reference/react-dom/static/prerender) and [`prerenderToNodeStream`](https://react.dev/reference/react-dom/static/prerenderToNodeStream) React APIs, which will enable you to server-side render, even if you are using suspense in your application.
Note that this is still only a step towards full support for streaming SSR, and you might need one of our [framework integrations](https://github.com/apollographql/apollo-client-integrations) to make full use of streaming SSR.

## TypeScript changes

### Moving types into namespaces

Most types are now colocated within the API that they relate to, so it is much easier to find the right type for the right situation.

Some examples:

- `FetchResult` is now `ApolloLink.Result`
- `ApolloClientOptions` is now `ApolloClient.Options`
- `QueryOptions` is now `ApolloClient.QueryOptions`
- `ApolloQueryResult` is now `ObservableQuery.Result`
- `QueryHookOptions` is now `useQuery.Options`
- `QueryResult` is now `useQuery.Result`
- `LazyQueryResult` is now `useLazyQuery.Result`

> These renames should be taken care of by running the codemod, but if you are not using the codemod, most of these types will still be available under their old names, with a deprecation note pointing to the new location.

Reviewer: ```suggestion
Many of the old, most-used types are still available in Apollo Client 4 (including the examples above), but are now deprecated in favor of the new namespaced types. We suggest running the codemod to update your code to use the new namespaced types.
```

Thoughts on reorganizing like this?

> ### Removal of `<TContext>` and `<TCacheShape>` generics

Reviewer: Can we split the `TCacheShape` to its own section? The intro paragraph only mentions `TContext` and provides good info about why we removed it, but it leaves out details on `TCacheShape`. That would help provide more info about `TCacheShape` on its own.

```suggestion
### Removal of `<TContext>` generic argument
```

> Many APIs in Apollo Client 3 used a manual `TContext` generic to pass the context type around. This was extremely fragile, so these generics have been removed in Apollo Client 4. Instead, override the `DefaultContext` type by using module augmentation.

Reviewer: ```suggestion
Many APIs in Apollo Client 3 provided a `TContext` generic argument to provide the context type. However, this generic argument was not consistently applied across all APIs that provided a `context` option and was easily forgotten. Apollo Client 4 removes the `TContext` generic argument in favor of using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) using the `DefaultContext` interface. This provides type safety throughout all Apollo Client APIs when using context.
```

(this assumes `TCacheShape` is moved to its own section). I used "easily forgotten" rather than "fragile" since I think the bulk of the problem is that users didn't define it at all in most cases. If you prefer "fragile", I'd suggest adding a bit more detail about what made it fragile.

To define types for your custom context properties, create a TypeScript file and define the `DefaultContext` interface.

```ts title="apollo-client.d.ts"
// This import is necessary to ensure all Apollo Client imports
// are still available to the rest of the application.
import "@apollo/client";

declare module "@apollo/client" {
  interface DefaultContext {
    myProperty?: string;
    requestId?: number;
  }
}
```

> Similar to `TContext`, the `TCacheShape` generic has been removed from the `ApolloClient` constructor options, because it added a lot of mental overhead with very little benefit. There is no replacement for this generic.

Reviewer: ````suggestion
### Removal of `<TCacheShape>`

The `TCacheShape` generic argument has been removed from the `ApolloClient` constructor. Most users set this to `any` which provided little benefit to type safety. APIs that previously relied on `TCacheShape` are now set to `unknown`.

To migrate, remove the `TCacheShape` generic argument when initializing `ApolloClient`.

```ts
new ApolloClient<any>({ // [!code --]
new ApolloClient({ // [!code ++]
  // ...
});
```
````

This can affect custom link implementations that rely on the `Observable` API.

If you were previously using the `map`, `filter`, `reduce`, `flatMap` or `concat` methods on the `Observable` instances, you will need to update your code to use the `rxjs` equivalents.

Additionally, if you were using `Obervable.of` or `Observable.from`, you will need to use the `rxjs` `of` or `from` functions instead.

```ts disableCopy=true
import { of, from, map } from "rxjs";// [!code ++]

Observable.of(1, 2, 3);// [!code --]
of(1, 2, 3);// [!code ++]
Observable.from([1, 2, 3]);// [!code --]
from([1, 2, 3]);// [!code ++]
observable.map((x) => x * 2);// [!code --]
observable.pipe(map((x) => x * 2));// [!code ++]
```

A good way to find the operator you need is to follow the [rxjs operator decision tree](https://rxjs.dev/operator-decision-tree).

## Local state changes

Local state management (using `@client` fields) has been removed from core and is now an opt-in feature. This change helps reduce the bundle size of your application if you don't use local state management features.

> To opt-in to use local state management, either by managing `@client` fields with the cache, or with local resolvers, initialize an instance of `LocalState` and provide it as the `localState` option to the `ApolloClient` constructor. Apollo Client will throw an error if a `@client` field is used and `localState` is not provided.

Author: ```suggestion
To opt-in to use local state management with the `@client` directive, initialize an instance of `LocalState` and provide it as the `localState` option to the `ApolloClient` constructor. Apollo Client will throw an error if a `@client` field is used and `localState` is not provided.
```

```ts
import { LocalState } from "@apollo/client/local-state";// [!code ++]

new ApolloClient({
  // ...
  localState: new LocalState(),// [!code ++]
  // ...
});
```

### Local resolvers

The `resolvers` option has been removed from the `ApolloClient` constructor and is now part of the `LocalState` class. Move local resolvers to the `LocalState` class.

```ts
new ApolloClient({
  resolvers: {...}, // [!code --]
  localState: new LocalState({
    resolvers: {...} // [!code ++]
  })
});
```

Now that local resolvers are part of the `LocalState` class, the following resolver methods were removed from `ApolloClient.prototype`:

- `addResolvers`
- `getResolvers`
- `setResolvers`

If you add resolvers after the initialization of the client, you may continue do so with your `LocalState` instance using the `addResolvers` method.

```ts
const client = new ApolloClient({
  // ...
});

client.addResolvers(resolvers);// [!code --]
client.localState.addResolvers(resolvers);// [!code ++]
```

> Stop using the `getResolvers` and `setResolvers` methods because they have not been ported to the `LocalState` class.

Author: ```suggestion
`getResolvers` and `setResolvers` have been removed without replacement. Stop using them.
```

```ts
const client = new ApolloClient({
  // ...
});

const resolvers = client.getResolvers();// [!code --]
client.setResolvers(resolvers);// [!code --]
```

#### Resolver context

> The context argument (i.e. the 3rd argument) passed to resolver functions has been updated. Apollo Client 3 spread [request `context`](../api/link/introduction#managing-context) and overrode the `client` and `cache` properties.

Author: ```suggestion
The context argument (i.e. the 3rd argument) passed to resolver functions has been updated. Apollo Client 3 spread [request `context`](../api/link/introduction#managing-context) and replaced any passed-in `client` and `cache` properties.
```

This example returns `undefined` due to the early return.

```ts
new LocalState({
  resolvers: {
    Query: {
      localField: () => {
        if (someCondition) {
          return;
        }

        return "value";
      },
    },
  },
});
```

This results in the following result:

```ts
{
  data: {
    localField: null,
  },
};
```

And a warning is emitted to the console:

```
The 'Query.localField' resolver returned `undefined` instead of a value.
This is likely a bug in the resolver. If you didn't mean to return a
value, return `null` instead.
```

Instead, the local resolver should return `null` instead.

```ts
if (someCondition) {
  return;// [!code --]
  return null;// [!code ++]
}
```

The exception is when running local resolvers to get values for exported variables for fields marked with the `@export` directive. For this case, You may return `undefined` to omit the variable from the outgoing request. If you need to know whether to return `undefined` or `null`, use the `phase` property provided to the context argument.

```ts
new LocalState({
  resolvers: {
    Query: {
      localField: (parent, args, { phase }) => {

>         // `phase` is "resolve" when resolving the field for the response

Author: ```suggestion
        // `phase` is "resolve" when resolving the field for the response // [!code highlight:4]
```

### Stage 3: Changes that may require refactoring

These changes may require code refactoring depending on your usage.

#### Changes affecting most users:

- **[Core API changes](#core-api-changes)** - Query behavior changes
- **[Error handling updates](#new-error-handling)** - New error classes and patterns
- **[TypeScript improvements](#typescript-changes)** - New type locations and patterns

#### Changes affecting React users:

- **[React hook changes](#for-react-users)** - `useLazyQuery`, `useMutation`, `useQuery` updates

#### Changes affecting fewer users:

- **[Link API changes](#link-api-changes)** - If you use custom links
- **[Local state changes](#local-state-changes)** - If you use `@client` directives
- **[Testing changes](#testing-related-changes)** - MockedProvider updates
- **[Bundling changes](#bundling-changes)** - New export patterns and transpilation
- **[Development mode changes](#development-mode-changes)** - Automatic detection

## Codemod

To ease the migration process, we have provided a codemod that automatically updates your codebase.

This codemod runs modifications in the following order:

<ol><li>

> 

Author: This is still missing the two latest additions to the codemod (compare codemod README)

Reviewer: This should now be updated

<tr><td>`__DEV__`</td><td>(unchanged)</td></tr>
</tbody><tbody>
<tr><th>`@apollo/client/utilities/global`</th><th>`@apollo/client/utilities/internal/globals`</th></tr>
<tr><td>`global`</td><td>(unchanged)</td></tr>
<tr><td>`maybe`</td><td>(unchanged)</td></tr>
</tbody><tbody>
<tr><th>`@apollo/client/utilities/global`</th><th>`@apollo/client/utilities/invariant`</th></tr>
<tr><td>`invariant`</td><td>(unchanged)</td></tr>
<tr><td>`InvariantError`</td><td>(unchanged)</td></tr>
<tr><td>`newInvariantError`</td><td>(unchanged)</td></tr>
</tbody>
</table>

</details>

### Replacing removed exports

If you used the [Codemod](#running-the-codemod) to update your imports, all removed exports moved to the `@apollo/client/v4-migration` entry point. You should get a TypeScript error everywhere you use a removed export. When you hover over the export, you'll get more information about why the export was removed along with migration instructions.

For a list of all removed imports and recommended actions, see [node_modules/@apollo/client/v4-migration.d.ts](https://app.unpkg.com/@apollo/client@^4.0.0-rc/files/v4-migration.d.ts)

## Update the initialization of `ApolloClient`

> Several of the constructor options of `ApolloClient` have changed in Apollo Client 4. This section provides instruction on migrating to the new options when initializing your `ApolloClient` instance.

Author: All of these are now handled by the codemod, assuming the intiializiation is not split up into something like
```ts
const options = {...}
const client = new ApolloClient(options)
```
that would *not* be handled.

So this should probably have a note here stating that most people should be good with the codemod.

```ts
import { Defer20220824Handler } from "@apollo/client/incremental"; // [!code ++]
const client = new ApolloClient({
  // ...
  incrementalHandler: new Defer20220824Handler(), // [!code ++]
  // ...
});
```

<Note>

Over time, we will introduce additional incremental handlers to support different proposals of the protocol.

</Note>

## Enable data masking types

In Apollo Client 4, the data masking types are now pluggable to allow for more flexibility when working with different solutions like GraphQL Code Generator or `gql.tadqa` - all of which might have different typings for data masking.

To configure the data masking types to be the same as they were in Apollo Client 3, create a `.d.ts` file in your project with the following content:

```ts title="apollo-client.d.ts
import { GraphQLCodegenDataMasking } from "@apollo/client/masking";

declare module "@apollo/client" {
  export interface TypeOverrides
    extends GraphQLCodegenDataMasking.TypeOverrides {}
}
```

## For React users

### (optional, but recommended) Stop using generated hooks

> Apollo Client 4 comes with a lot of improvements to our hook types. If you are using generated hooks, you will not be able to benefit from these improvements, so we recommend to stop using generated hooks and use our hooks directly with `TypedDocumentNode` instead.

Author: Should we link to the guilds codemod video here? It shows migration to the client preset, so that might need additional explanation/a link to the according section in our types docs.

{/* prettier-ignore */}
```ts disableCopy=true
import { CombinedProtocolErrors } from "@apollo/client"; // [!code ++]

const { error } = useQuery(QUERY);

if (error.protocolErrors) { // [!code --]
  error.graphQLErrors.map((graphQLError) => {/* ... */}); // [!code --]
if (CombinedProtocolErrors.is(error)) { // [!code ++]
  error.errors.map((graphQLError) => {/* ... */}); // [!code ++]
}
```

#### Migrate from `clientErrors`

The `clientErrors` property was not used by Apollo Client and therefore has no replacement. Any non-GraphQL errors or non-protocol errors are passed through as-is.

### Errors as guaranteed error-like objects

Apollo Client 4 guarantees that the `error` property is an `ErrorLike` object, an object with a `message` and `name` property. This avoids the need to check the type of `error` before consuming it. To make such a guarantee, thrown non-error types are wrapped.

> Thrown values that are non-error-like objects (such as a symbol or primitive value) are wrapped in an `UnconventionalError` error instance to avoid the need to check the type of the error before consuming it. `UnconventionalError` sets its `name` to `UnconventionalError`, sets its `message` to `"An error of unexpected shape occured."`, and sets its [`cause` property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) to the value originally thrown.

Reviewer: [476d59f](https://github.com/apollographql/apollo-client/pull/12836/commits/476d59f59066fc4cb7ba31370903e38d8fee5efe)
